WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,89 +35,154 @@ private class WeeklyRoundupDataProvider {
+    ///
+    /// This function fetches the stats for a single site and passes the result to a completion handler.
+    /// If it encounters any error during fetching, it calls the completion handler with an appropriate error object.
+    /// The completion handler is executed in the same queue as the provided `NSManagedObjectContext`.
+    ///
+    /// **Note:** The completion handler is executed on the context's queue.
+    private func fetchStats(
+        for site: Blog,
+        endingOn periodEndDate: Date,
+        authToken: String,
+        in context: NSManagedObjectContext,
+        completion: @escaping (Result< StatsSummaryData?, Error>) -> Void
+    ) {
+        guard let dotComID = site.dotComID?.intValue else {
+            completion(.failure(DataRequestError.dotComSiteWithoutDotComID(site)))
+            return
+        }
+        let wpApi = WordPressComRestApi.defaultApi(oAuthToken: authToken, userAgent: WPUserAgent.wordPress())
+        let statsServiceRemote = StatsServiceRemoteV2(wordPressComRestApi: wpApi, siteID: dotComID, siteTimezone: site.timeZone)
+        statsServiceRemote.getData(for: .week, endingOn: periodEndDate, limit: 1) { (timeStats: StatsSummaryTimeIntervalData?, error) in
+            context.perform {
                 guard let timeStats = timeStats else {
-                    guard let error = error else {
-                        self.onError(DataRequestError.unknownErrorRetrievingStats(site))
-                        return
+                    if let error = error {
+                        completion(.failure(DataRequestError.errorRetrievingStats(dotComID, error: error)))
+                    } else {
+                        completion(.failure(DataRequestError.unknownErrorRetrievingStats(site)))
                     }
-
-                    self.onError(DataRequestError.errorRetrievingStats(dotComID, error: error))
-                    return
-                }
-
-                guard let stats = timeStats.summaryData.first else {
-                    // No stats for this site, or not enough views to qualify.  This is not an error.
                     return
                 }
-
-                blogStats[site] = stats
+                completion(.success(timeStats.summaryData.first))
             }
         }
     }
 
+    /// Returns the end date for the period (the start of the current week).
+    /// This function creates a calendar with a GMT timezone and specifies that it needs a date representing the start of the current week.
+    /// If it can't find a valid date, it returns nil.
+    private func getPeriodEndDate() -> Date? {
+        var gmtCalendar = Calendar(identifier: .gregorian)
+        gmtCalendar.timeZone = TimeZone(secondsFromGMT: 0)!
+        var endDateComponents = DateComponents()
+        endDateComponents.weekday = 1
+        return gmtCalendar.nextDate(after: Date(), matching: endDateComponents, matchingPolicy: .nextTime, direction: .backward)
```

Comment:
This is an existing logic and I didn't change anything here. I just extracted this logic from `getTopSiteStats` into its own method.